### ISSUE TITLE
Ridiculously big CPU and RAM usage

### ISSUE THREAD

**Issue by zetok** [qTox/qTox#266]
![2014-09-08 qtox ram usage messages](https://cloud.githubusercontent.com/assets/3148759/4189135/442c05d0-3779-11e4-89e4-aa12e7628e4f.png)

Apparently qTox doesn't handle well receiving a lot of messages.

I've sent thousands of lines of text to qTox, probably not more than 100k.

During, and after qTox remained hanged.

Tested on https://github.com/tux3/qTox/commit/cfb3313821e28a3f40c9712c5f7b675f796e5e7f

**Comment by tux3** [comment]
That probably has to do with the layout/widget spam, it'll be fixed Soon™

**Comment by zetok** [comment]
After #264 got merged RAM usage dropped significantly (only few MB increase during test), but unfortunately CPU usage stays the same, and after receiving 10k lines sample qTox stays hanged.

**Comment by tux3** [comment]
Do you use some sort of script I could use to send the tens of thousand of lines ?

It should be easy to run that in a profiler.

**Comment by zetok** [comment]
Nah, I'm just using Toxic.

Copy content of some big file and paste into Toxic. Since new versions support faux offline messaging and read recipes, one knows whether messages were really received.

**Comment by tux3** [comment]
Profiling shows that there is a massive slowdown in chatareawidget.cpp, updateChatContent()
The call to setHtml() alone is responsible for >60% of the CPU used by qTox over the whole duration of the test.

**Comment by tux3** [comment]
Please comment if you still have this bug and I'll reopen. 
If you notice any strange behavior of the qTox chat after the commit that closed this issue, please also report it since I was particularly tired when I wrote the fix and can't exclude the possibility that I may have introduced new bugs.

**Comment by zetok** [comment]
Situation improved, after sending 10k lines of text qTox only has increase of RAM usage up to 419 MB. While receiving for a short moment became unresponsive.

I'm going to test how well it'll behave with 10 million lines sample..

**Comment by dubslow** [comment]
10K lines of text taking up 419MB is still rather a lot...

**Comment by tux3** [comment]
That last patch kind of broke everything. I'll make a real fix.

**Comment by dubslow** [comment]
I think the issue is [getHtmledMessages](https://github.com/tux3/qTox/blob/737e76c72fdd966014744b17b80360d9e6258c71/widget/chatareawidget.cpp#L74). Everytime it's called (i.e. everytime updateChatContent is called), it constructs the entire chat history via message-by-message string addition. 

With 10K messages, you are adding to a string in place 10K times -- lots of CPU usage and malloc calls. For every updateChatContent call.

Either use [QStringBuilder](http://qt-project.org/doc/qt-5/qstring.html#more-efficient-string-construction) or store the entire html string instead of the Action objects. (The latter solution might help with RAM -- though maybe the Action objects are still required for some other purpose? Maybe some hybrid solution?)

**Comment by tux3** [comment]
Exactly. I'm trying to use QTextCursor instead.

getHtmledMessages itself isn't the slow part (source: profiling), but setHtml() uses a LOT of CPU and "messages" uses a lot of ram since it has a copy of everything.
Additionnally, actions that don't need to update themselves ever can just discard their QTextCursor to save some RAM.

The cursors would be direct references to Qt's internal copy of the HTML, and each message/file transfert widget/... could update itself through it's cursor without having to redraw the whole chatbox.

**Comment by zetok** [comment]
It's not fixed.
Change made things worse, now newest qTox is instantly DoS-able (hangs) when someone is sending a lot of messages.

Only improvement I see is RAM usage, which isn't that important if client refuses to work at all.

**Comment by tux3** [comment]
TODO: threads.

**Comment by zetok** [comment]
Threads will help, but not much since CPU usage is 
![2014-09-10 qtox long messages not displayed correctly](https://cloud.githubusercontent.com/assets/3148759/4253048/7edd8968-3a9b-11e4-84e8-6d36d94abff8.png)

**Comment by tux3** [comment]
Profiling says :

<blockquote><pre>
# ChatAreaWidget::insertMessage        8,200,000,000 units
## QTextCursor::insertHtml             3,400,000,000 units
## QTextCursor::setBlockFormat         3,200,000,000 units
## QTextCursor::appendRows             1,700,000,000 units
## &lt;rest is negligible>
</pre></blockquote>

So all the slowness is in Qt itself, and unless we can replace setBlockFormat by something less expensive I can't do anything about CPU usage, since the chatform is HTML now, Qt has a lot of job to do to render that correctly.

What I can do is keep the UI responsive and show the messages as they are processed, a couple of seconds late if they are really heavy.

**Comment by apprb** [comment]
Unfortunately haven't used any profiler before (and it is not the best task I can do at 2 AM), so I can't clearly say how it helps.

**Comment by tux3** [comment]
Actually now that I think about it, threading this is going to be hard / impossible since the paint events could try to access the ChatAreaWidget at any time.

I'm not sure how to solve this. That said, if one of your contact tries really hard to DoS your client, you can always remove that contact. I don't think it'll be possible to completely prevent someone in your contact list to DoS your client if they want to.

I'll just add some calls to processEvents() in the meantime.

**Comment by tux3** [comment]
Alright, with the new commits qTox is much faster.

<blockquote><pre>
# ChatAreaWidget::insertMessage        4,200,000,000 units
## QTextCursor::insertHtml             1,500,000,000 units
## QTextCursor::insertText             1,300,000,000 units
## QTextCursor::appendRows             1,300,000,000 units
## &lt;rest is negligible>
</pre></blockquote>

**Comment by zetok** [comment]
I haven't sent any text:
![2014-09-10 qtox long messages not displayed correctly](https://cloud.githubusercontent.com/assets/3148759/4408256/9e69ee3a-44d1-11e4-85b2-625228ce8190.png)

Instead I've tried file transfers. 18/18 failed made it.

Edit: I was sending files from qTox

**Comment by zetok** [comment]
Receiving ~50 MB file increases RAM usage by ~300 MB, and during file transfer qTox is hanged.

**Comment by dubslow** [comment]
I'm reconfirming here that input lag in chat forms scales with the total chat history. Total history is roughly ten full screens worth at 1080p, and chat delay of more than half a second, maybe a full second.

I just counted roughly 40 lines per screen, so say ~400 lines over several hours. I know for a fact that Skype can scroll through several thousand lines of history, so...

Edit: If I may, perhaps we should review the code of outside projects. [Quassel](https://github.com/quassel/quassel/) is a fully featured IRC client written in Qt (and is not coincidentally the one I use). Maybe we could even ask for help or advice. (Among other things, although you can copy qTox chats, the format in which they paste is still godawful.)

https://github.com/quassel/quassel/tree/master/src/qtui

**Comment by krepa098** [comment]
Seems like building a chat-log on top of QGraphicsView is the way to go. I did a little prototype and it works quite well. The downside of using QGraphicsView is that you have to re-implement text selection behavior and stuff like that. On the other hand you can insert whatever type of content you want (pics, controls, videos, websites...) and it's always instantaneous no matter how many lines of text there are.

**Comment by apprb** [comment]
Before PR #395 problem problem was caused by redrawing whole chat on every new message insertion (my guess about the problem origin), in other words every message depends on all other messages, so insertion become no better that O(n). Now I believe CPU usage problem is somehow resolved.
Can you share somewhere your prototype if it isn't hard for you?

**Comment by krepa098** [comment]
> Can you share somewhere your prototype if it isn't hard for you?

Yes if it turns out to be useful (it's quite a hacky mess right now).

**Comment by dubslow** [comment]
I do believe the current implementation needs fixing; if nothing else, even though text is copyable, pasting across table cell boundaries causes problems.

**Comment by fcore117** [comment]
i hope this huge ram issue will be fixed, i got even 210mb sometimes. 9-15mb would be optimal.

**Comment by krepa098** [comment]
Alright I'm finally getting somewhere :)
![snapshot23](https://cloud.githubusercontent.com/assets/4587853/4827481/db3d28f6-5f78-11e4-929b-625d609d5a62.png)

Memory consumption is about 40MB at 5000 messages. I think that's pretty reasonable.

**Comment by apprb** [comment]
Good job!! 
And it looks like you somehow solved filetransfer widget's resize/content fitting problem.

**Comment by fcore117** [comment]
today lastest release
![qtoxmem](https://cloud.githubusercontent.com/assets/1313813/4981532/eb76e622-6909-11e4-808f-837438130f29.png)

**Comment by zetok** [comment]
Yep, have to agree with @fcore117 , on qTox 9e7ee848e88dd359224f55a6c066ea1819125268:
![invite](https://cloud.githubusercontent.com/assets/3148759/4983334/c8907f64-6919-11e4-8e11-902adc803fb9.png)

Increase in used memory was rather slow, but during ~2h it increased ~1.5GB. Screenshot done while RAM usage was still rising, perhaps I'll put new screenshot when it'll reach some nice value, or I'll drop a comment when it finally will trigger OOM killer.

I'm not sure what exactly caused it.
So, since groupchat invites usually don't work for me (#394), I've setup Toxic to invite my qTox instances to groupchat by spamming qTox with 2456 × `/invite 0`, which should surely invite qTox to groupchat.

For some reason that didn't seem to work. Upon issuing just single `/invite 0` Toxic returned:
`Failed to invite contact to group.`
At that time last log entries contained:

```
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Trying to join groupchat invite by friend 7"
"Core: Group invite by 7"
"Trying to join groupchat invite by friend 7"
"Core: Group invite by 7"
"Trying to join groupchat invite by friend 7"
"Core: Group invite by 7"
"Trying to join groupchat invite by friend 7"
"Core: Group invite by 7"
"Trying to join groupchat invite by friend 7"
"Core: Group invite by 7"
"Trying to join groupchat invite by friend 7"
```

After that issuing new invites didn't seem to produce any log. The only thing that later on appeared was:

```
Widget: auto away activated
Settings: Saving in  "/home/zetok/.config/tox/settings.ini"
Core: writing tox_save to  "/home/zetok/.config/tox/_3.tox"
Widget: auto away deactivated
Settings: Saving in  "/home/zetok/.config/tox/settings.ini"
Core: writing tox_save to  "/home/zetok/.config/tox/_3.tox"
Widget: auto away activated
Settings: Saving in  "/home/zetok/.config/tox/settings.ini"
Core: writing tox_save to  "/home/zetok/.config/tox/_3.tox"
```

Which obviously has nothing to do with experienced issues

Whole libtoxcore.log of session:

```
============== Starting logger [6674] ==============
Time: Mon Nov 10 18:10:36 2014

[6674] [DEBUG] [0] [network.c:570 new_networking()] Dual-stack socket: enabled
[6674] [DEBUG] [0] [network.c:584 new_networking()] Local multicast group FF02::1 joined successfully
[6674] [DEBUG] [0] [network.c:613 new_networking()] Bound successfully to [::]:33445
[6674] [DEBUG] [0] [msi.c:1584 msi_init_session()] New msi session: 0x7f0cf0004020 max calls: 16
[6674] [DEBUG] [1690574] [DHT.c:369 client_or_ip_port_in_list()] coipil[4]: switching client_id (ipv4)
[6674] [DEBUG] [3750577] [DHT.c:369 client_or_ip_port_in_list()] coipil[15]: switching client_id (ipv4)
[6674] [DEBUG] [3978597] [DHT.c:369 client_or_ip_port_in_list()] coipil[15]: switching client_id (ipv4)
[6674] [DEBUG] [4230631] [DHT.c:369 client_or_ip_port_in_list()] coipil[15]: switching client_id (ipv4)
[6674] [DEBUG] [4890623] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [5418624] [DHT.c:369 client_or_ip_port_in_list()] coipil[7]: switching client_id (ipv4)
[6674] [DEBUG] [5678644] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [6066646] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [6451051] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [6658853] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [7130583] [DHT.c:369 client_or_ip_port_in_list()] coipil[23]: switching client_id (ipv4)
[6674] [DEBUG] [7750680] [DHT.c:369 client_or_ip_port_in_list()] coipil[14]: switching client_id (ipv4)
[6674] [DEBUG] [7834843] [DHT.c:369 client_or_ip_port_in_list()] coipil[7]: switching client_id (ipv4)
```

I've spammed qTox a bit more with invites, not really hoping that it would help with anything.

Then I've proceeded to testing how well qTox would behave after that.

For that I've copied `Failed to invite contact to group.` and tried to send it to qTox multiple (hundreds?)  times. What was quite astonishing, qTox seemed to receive those messages, but didn't seem to return read receipts correctly, which caused Toxic to believe that messages weren't delivered correctly, thus causing Toxic to send those messages over and over again.

To add more fun to that, both clients seemed to see each other going offline every few seconds, whereas 3rd client didn't see qTox or Toxic going offline.

To confirm that it wasn't Toxic at fault with messages, I've sent few messages from Toxic to other qTox instance, and no problems with read recipes occurred. Though I haven't spammed messages, so…

I haven't tried to reproduce it.

**Comment by fcore117** [comment]
zetok: my increase was only after some 4-9 minutes.

**Comment by dubslow** [comment]
The increase happens without any messaging activity?

zetok: qTox isn't receiving any peer lists from whoever invites you. I'll try and make a core patch with some more logging statements.

**Comment by zetok** [comment]
```
lis 11 09:31:33 galaxor kernel: Out of memory: Kill process 6674 (qtox) score 1790 or sacrifice child
lis 11 09:31:33 galaxor kernel: Killed process 6674 (qtox) total-vm:12308716kB, anon-rss:4279384kB, file-rss:17708kB
```

:3

**Comment by tux3** [comment]
<p>>12 GB used</p>

**Comment by ghost** [comment]
> having more than 8 GB of ram

why (save the running qTox)?

**Comment by zetok** [comment]
qTox d87cf99d8fb786ba102b7ac86e96c0dad9a12523
After some time in audio groupchat:
![agroupchat](https://cloud.githubusercontent.com/assets/3148759/5022392/4a7d8bd4-6adc-11e4-8801-db387c4ea0b6.png)

I don't really know what caused it, and whether reproducible, but I'm going to wildly guess that there's memleak somewhere, since I haven't done anything beside being in audio groupchat

**Comment by dubslow** [comment]
No messages from any friends or whatever?

**Comment by zetok** [comment]
No, the only thing I did beside joining groupchat was to switch few times status, and set (again) alias `stqism`. I've left qTox running overnight, and even though it stayed in groupchat, there was only slight increase (5 MB) in RAM usage. Beside that overnight status again was changed from "Busy" to "Online"..

**Comment by zetok** [comment]
![memleak](https://cloud.githubusercontent.com/assets/3148759/5034027/0a5f343a-6b63-11e4-9d95-7d2188ab35c7.png)
Same qTox instance, it definitely has something to do with groupchats.

**Comment by fcore117** [comment]
having 11 contacts in list increases memory too somewhere 140mb, with blank account it uses some 75mb

**Comment by dubslow** [comment]
I've been in the group chat for a couple of days, no audio, and everything seems fine. Note that for the last several hours I was left on my own in the chat because of the split or whatever it was that happened.

```
  PID USER      PR  NI  VIRT  RES  SHR S  %CPU %MEM     TIME+ COMMAND
 4982 bill      20   0  817m 255m 127m S  23.3  1.6   2345:11 chrome            
 4937 bill      20   0 1529m 305m 108m S   4.0  1.9 314:54.92 chrome            
 5138 bill      20   0 1529m 253m  40m S   4.0  1.6  60:53.70 chrome            
12317 bill      20   0  534m 175m  67m S   3.7  1.1 110:58.31 steam             
 4223 bill       9 -11  717m  11m 6104 S   3.0  0.1 122:48.41 pulseaudio        
 4414 bill      20   0 2253m 558m  74m S   2.7  3.5 247:12.91 cinnamon          
24261 bill      20   0  522m 146m  41m S   2.3  0.9 168:03.25 skype             
 5678 bill      20   0 1323m 114m  51m S   2.0  0.7  54:09.18 qtox              
 3760 root      20   0  354m 140m  67m S   1.3  0.9 124:13.02 Xorg              
 4441 bill      20   0 2472m 760m  80m S   1.0  4.7 121:01.06 bitcoin-qt        
 5028 bill      20   0 1028m 250m  27m S   0.7  1.6  56:21.31 chrome            
12432 bill      20   0  547m  45m  23m S   0.7  0.3   1:25.63 gnome-terminal-   
  701 bill      20   0 1101m 169m  640 S   0.3  1.1   8:25.17 Chrome_ProcessL   
 5018 bill      20   0  728m  28m  688 S   0.3  0.2  10:33.19 Chrome_ProcessL  
```

Note that qTox is about on par with Skype, though it's virtual memory usage is substantially more.

**Comment by zetok** [comment]
Just sending files from qTox resulted in:

```
Dec 29 01:07:00 localhost kernel: Out of memory: Kill process 5426 (qtox) score 303 or sacrifice child
Dec 29 01:07:00 localhost kernel: Killed process 5426 (qtox) total-vm:2770492kB, anon-rss:1836348kB, file-rss:8096kB
```

qTox sent successfully 216 files, in total they were 1.3 GB big. At that point qTox already used well >2 GB of RAM.
Trying to send another 27 files resulted in OOM killer being triggered. Those files were 182.3 MB big.

**Comment by Nokta-strigo** [comment]
Tried to send about 200 of photos (around 4 MB every) in one portion and had to kill receiving qTox when it's memory usage became > 4 GB.
May be it's better to transfer only 1-3 files simultaneously and put other files in a queue? It would be even better in terms of writing it to HDD.

**Comment by zetok** [comment]
@Nokta-strigo having a hard limit on how many files can be sent simultaneously is generally a bad idea.

What could be done though, is to provide an option to send `x` files simultaneously while rest is put in queue, with some sane default like `3`, with values ranging from 1 to 256 files (since AFAIK current toxcore API allows up to 256 file transfers, probably will(?) be changed in new API).

Anyway, it doesn't change the fact that qTox poorly behaves with managing RAM usage when it comes to file transfers.

**Comment by Nokta-strigo** [comment]
Yes, that's what I was talking about, and flexibility is surely better!

**Comment by dubslow** [comment]
Closeable?

**Comment by agilob** [comment]
Bump.

**Comment by tux3** [comment]
Fixed by the recent memory leak/saving patches and new API qTox.

It's probably still possible to choke qTox if you try really hard, but it's not ridiculous anymore.

**Comment by zetok** [comment]
![nope](https://cloud.githubusercontent.com/assets/3148759/7728336/93f68ffe-ff06-11e4-91a3-0c0b8ae173dc.png)
qTox 4306f87df4a0825a19b3a2d0cd4e4b99545acf36 ↑

CPU usage was up to 150%.

Doesn't seem to be fixed. Please reopen.

**Comment by agilob** [comment]
How do you test it? I never get more than 10%

**Comment by zetok** [comment]
I've started sending 2 big files (11 GB each) to one contact on localhost. Sadly because of #1693 it didn't work too well to make qTox misbehave, but I at least was getting ~19 MB/s and CPU usage ~50%.

While files were being sent to 1st contact., I've started sending bunch of small files to other contact on localhost (by setting "Receive files automatically" and "drag & drop" files). That's where problems started, CPU went up to ~150%, and ram up to 1.6 GB. After adding more files file transfer with first contact broke (wasn't resumed, but that's probably not a bug, since I was receiving to `/dev/null`), and there were additionally some graphical glitches (#1694)

**Comment by tux3** [comment]
@zetok File transfers taking more CPU is by design, otherwise with all the toxcore overhead we could never reach decent transfer speeds.

Howether I'm not sure what exactly is the scenario where you get 150% usage. How exactly, step by step, do you do that? How long does it last (5s, 5 minutes, 2 hours)? Is it only with many small files?

**Comment by agilob** [comment]
> File transfers taking more CPU is by design, otherwise with all the toxcore overhead we could never reach decent transfer speeds.

@zetok We had a talk about it in context of toxvpn, remember?

**Comment by zetok** [comment]
@tux3 

> Howether I'm not sure what exactly is the scenario where you get 150% usage. How exactly, step by step, do you do that? How long does it last (5s, 5 minutes, 2 hours)? Is it only with many small files?

It's about transferring a lot of graphical files - for each there has to be made thumbnail, which adds ~100% of CPU usage to ~50% of CPU usage that is already there due to big files being transferred, and it lasts until thumbnails are generated.

**Comment by tux3** [comment]
@zetok So the problem is that you transferred huge image files, and it took a while to generate the thumbnails?
Can you send me a link to where I can get an image like that for reference?

**Comment by zetok** [comment]
@tux3 not quite, I've used bunch of really small images.. There's a sample one, confirmed to "work" for me: https://a.pomf.se/viodwi.tar.xz

They all need to be dragged & dropped into qTox. When doing it while other, large file transfer is taking place already:
![fu fu fu](https://cloud.githubusercontent.com/assets/3148759/7738242/27ce12f4-ff4a-11e4-89b9-37edca641cbb.png)

**Comment by anoadragon453** [comment]
Is this still an issue with latest qTox?

**Comment by zetok** [comment]
Yep, on 2d06b9960c60f1fc886ae04c595bd9a921cff316:
![muh files](https://cloud.githubusercontent.com/assets/3148759/16615318/c76f6c46-436d-11e6-9a32-a475b19fc3f6.png)
↑ top one is sending, bottom one is receiving (automatically accepts FT), done with a sample of 8k images (didn't send them all, just started sending, and turned qTox off, since ram usage skyrocketed)

UI of sending qTox froze, receiving one worked, but slowly.

**Comment by zetok** [comment]
Though this probably should be moved to new issue, and this one closed, since it's rather messy by now.

**Comment by Geremia** [comment]
I don't get high CPU percentage, but my qTox's CPU time is higher than any other of my processes (even higher than those running 24/7, like my web server).

**Comment by Diadlo** [comment]
Is it still relevant? If so, let's reopen as @zetok suggested

**Comment by sudden6** [comment]
No response
